### ISSUE TITLE
*_if operates on all columns if a predicate returns just character() (= no match)

### ISSUE THREAD

**Issue by emailmarketingtips** [hadley/dplyr#1989]
I don't know if this is an intended behavior (confused me): 

```
library(dplyr)
df <- data_frame( x = as.factor("a"), y = 1L )
str(df)
 # $ x: Factor w/ 1 level "a": 1
 # $ y: int 1

dplyr:::probe_colwise_names(df, is.character)
# character(0)

df <- mutate_if(df, is.character, as.factor)
str(df)
# Classes ‘tbl_df’, ‘tbl’ and 'data.frame':	1 obs. of  2 variables:
#  $ x: Factor w/ 1 level "a": 1
#  $ y: Factor w/ 1 level "1": 1 #
```
`mutate_if` turned  the integer column `y` into a factor on an is.character check. 

My `packageVersion("dplyr")` is `‘0.5.0.9000’`

**Comment by karimn** [comment]
Is this fix not in dplyr v0.5.0? I'm seeing this:
```
> data.frame(x = 1:10, y = 1:10) %>% dplyr::mutate_if(is.factor, funs(fct_relevel(., "10"))) 
Error in mutate_impl(.data, dots) : 
  `f` must be a factor (or character vector).
```

**Comment by yutannihilation** [comment]
@karimn 
It seems that v0.5.0 does not contain the fix (as it was committed in Aug 2016, whereas v0.5.0 was released in June 2016). You need to `install_github("hadley/dplyr")` for now.
